velox/benchmarks/tpch/TpchBenchmark.cpp

```diff
@@ -0,0 +1,148 @@
+}
+static bool validateFormat(const char* flagname, const std::string& value) {
+  if ((value.compare("parquet") == 0) ||
+      (value.compare("orc") == 0)) // value is ok
+    return true;
+  std::cout
+      << fmt::format(
+             "Invalid value for --{}: {}. Allowed values are [\"parquet\", \"orc\"]",
+             flagname,
+             value)
+      << std::endl;
+  return false;
+}
+} // namespace
+
+DEFINE_string(data_path, "", "Root path of TPC-H data");
+DEFINE_int32(
+    run_query_verbose,
+    -1,
+    "Run a given query and print execution statistics ");
+DEFINE_int32(num_drivers, 4, "Number of drivers");
+DEFINE_string(format, "parquet", "Data format");
+DEFINE_int32(num_splits_per_file, 10, "Number of splits per file");
+
+DEFINE_validator(data_path, &notEmpty);
+DEFINE_validator(format, &validateFormat);
+
+class TpchBenchmark {
+ public:
+  void initialize() {
+    functions::prestosql::registerAllScalarFunctions();
+    filesystems::registerLocalFileSystem();
+    parquet::registerParquetReaderFactory();
+    dwrf::registerDwrfReaderFactory();
+    auto hiveConnector =
+        connector::getConnectorFactory(
+            connector::hive::HiveConnectorFactory::kHiveConnectorName)
+            ->newConnector(kHiveConnectorId, nullptr);
+    connector::registerConnector(hiveConnector);
+  }
+
+  std::shared_ptr<Task> run(const TpchPlan& tpchPlan) {
+    CursorParameters params;
+    params.maxDrivers = FLAGS_num_drivers;
+    params.numResultDrivers = 1;
+    params.planNode = tpchPlan.plan;
+    const int kNumSplitsPerFile = FLAGS_num_splits_per_file;
+
+    bool noMoreSplits = false;
+    auto addSplits = [&](exec::Task* task) {
+      if (!noMoreSplits) {
+        for (const auto entry : tpchPlan.dataFiles) {
+          for (const auto path : entry.second) {
+            auto const splits = HiveConnectorTestBase::makeHiveConnectorSplits(
+                path, kNumSplitsPerFile, tpchPlan.dataFileFormat);
+            for (const auto& split : splits) {
+              task->addSplit(entry.first, exec::Split(split));
+            }
+          }
+          task->noMoreSplits(entry.first);
+        }
+      }
+      noMoreSplits = true;
+    };
+    auto [cursor, results] = readCursor(params, addSplits);
+    return cursor->task();
+  }
+};
+
+TpchBenchmark benchmark;
+std::shared_ptr<TpchQueryBuilder> queryBuilder;
+
+BENCHMARK(q1) {
+  const auto planContext = queryBuilder->getQueryPlan(1);
+  benchmark.run(planContext);
+}
+
+BENCHMARK(q6) {
+  const auto planContext = queryBuilder->getQueryPlan(6);
+  benchmark.run(planContext);
+}
+
+BENCHMARK(q18) {
+  const auto planContext = queryBuilder->getQueryPlan(18);
+  benchmark.run(planContext);
+}
+
+int main(int argc, char** argv) {
+  folly::init(&argc, &argv, false);
+  benchmark.initialize();
+  queryBuilder = std::make_shared<TpchQueryBuilder>(toFileFormat(FLAGS_format));
+  queryBuilder->initialize(FLAGS_data_path);
+  if (FLAGS_run_query_verbose == -1) {
```

Comment:
This flag is confusing to me. The name suggests that it allows to restrict the benchmark to just 1 query and also print out some extra information about the run. The actual meaning is different though.

Reply:
@mbasmanova Can you please elaborate on "The actual meaning is different though."? I checked the "folly benchmark" API and there is no way to specify a particular benchmark to run. So the code here manually invokes the `benchmark.run()`.

Reply:
My understanding that running a benchmark via folly::runBenchmarks(); includes running the code multiple times measuring the execution time and reporting some average or min of these time. When run_query_verbose is enabled, we do not do that. Instead, we run the code just once and report the execution time from a single run. Please, correct me if my understanding is off.

Reply:
Your understanding is correct. `run_query_verbose` only executes the query once. The folly benchmark API does not allow us to tap into the execution for a single benchmark.

Reply:
As the `run_query_verbose` flag mentions `query` and not `benchmark` this should be okay?